network/resource_use_stats.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+package network
+
+import (
+	"sort"
+	"strconv"
+	"sync"
+
+	"github.com/ethersphere/swarm/log"
+)
+
+// resourceUseStats can be used to count uses of resources. A Resource is anything with a Key()
+type resourceUseStats struct {
+	resourceUses map[string]int
+	waiting      map[string]chan struct{}
+	lock         sync.RWMutex
+	quitC        chan struct{}
+}
+
+// Resource represents anything with a Key that can be accounted with some stat.
+type Resource interface {
+	Key() string   // unique id in string format of the resource.
+	Label() string // short string format of the key for debugging purposes.
+}
+
+type ResourceCount struct {
+	resource Resource
+	count    int
+}
+
+func newResourceUseStats(quitC chan struct{}) *resourceUseStats {
```

Comment:
This should be a read only chan, to be explicit that is in not closed or written to by resourceUseStats.

Reply:
Changed to readOnly